### PR TITLE
Rc 2.5.0

### PR DESCRIPTION
**Pull request by tomazpu**

### CODE REVIEW

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 25
- Added 2
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/models/PaymentSofort.php  1
- wirecardpaymentgateway/classes/Response/Failure.php  1
- wirecardpaymentgateway/helper/StringHelper.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/EntityBuilderFactory.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/Basket/BasketBuilder.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItem.php  3
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/PostProcessingTransactionBuilder.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/InitialTransactionBuilder.php  1
- wirecardpaymentgateway/helper/PaymentProvider.php  2
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/helper/ArrayHelper.php  3
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/TransactionCart.php  1
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItemCollection.php  3
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -1
+ wirecardpaymentgateway/controllers/admin/WirecardTransactions.php  -10
+ wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php  -3
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         

Clones removed
==============
+ wirecardpaymentgateway/models/PaymentSepaDirectDebit.php  -1
+ wirecardpaymentgateway/models/PaymentAlipayCrossborder.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405&bid=15630647)

File: wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    private function isAutoCaptureEnabled()
+    {
+        return intval(Configuration::get(ConfigConstants::SETTING_GENERAL_AUTOMATIC_CAPTURE_ENABLED));
+    }
+
+    /**
+     * @return int
+     * @since 2.5.0
+     */
+    private function isForceOrderStateChangeAllowed()
```
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php]: Codacy Issue found: [Avoid unused private methods such as 'isForceOrderStateChangeAllowed'.](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

File: wirecardpaymentgateway/classes/Constants/FormConstants.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * @return array
+     * @since 2.5.0
+     */
+    public static function getElementTypesWithValues()
+    {
+        return [
+            self::FORM_ELEMENT_TYPE_SWITCH
+        ];
+    }
+}
```
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Constants/FormConstants.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

File: wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php (hunk 1 above)
Comment (tomazpu) [reply wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php]: Please remove this method before release

File: wirecardpaymentgateway/classes/Constants/FormConstants.php (hunk 2 above)
Comment (tomazpu) [reply wirecardpaymentgateway/classes/Constants/FormConstants.php]: This is a false positive from codacy.

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 25
- Added 2
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/TransactionCart.php  1
- wirecardpaymentgateway/models/PaymentSofort.php  1
- wirecardpaymentgateway/classes/Response/Failure.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/helper/StringHelper.php  1
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItemCollection.php  3
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/EntityBuilderFactory.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/InitialTransactionBuilder.php  1
- wirecardpaymentgateway/helper/PaymentProvider.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/Basket/BasketBuilder.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItem.php  3
- wirecardpaymentgateway/helper/ArrayHelper.php  3
- wirecardpaymentgateway/classes/Transaction/Builder/PostProcessingTransactionBuilder.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -1
+ wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php  -3
+ wirecardpaymentgateway/controllers/admin/WirecardTransactions.php  -10
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         

Clones removed
==============
+ wirecardpaymentgateway/models/PaymentAlipayCrossborder.php  -1
+ wirecardpaymentgateway/models/PaymentSepaDirectDebit.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405&bid=15630647)

File: wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php (hunk 1 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php]: Codacy Issue found: [Avoid unused private methods such as 'isForceOrderStateChangeAllowed'.](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

File: wirecardpaymentgateway/classes/Constants/FormConstants.php (hunk 2 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Constants/FormConstants.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 25
- Added 2
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Transaction/Builder/InitialTransactionBuilder.php  1
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/Basket/BasketBuilder.php  2
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/helper/ArrayHelper.php  3
- wirecardpaymentgateway/models/PaymentSofort.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItemCollection.php  3
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/helper/StringHelper.php  1
- wirecardpaymentgateway/helper/PaymentProvider.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/TransactionCart.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/PostProcessingTransactionBuilder.php  2
- wirecardpaymentgateway/classes/Response/Failure.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/EntityBuilderFactory.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItem.php  3
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -1
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php  -3
+ wirecardpaymentgateway/controllers/admin/WirecardTransactions.php  -10
         

Clones removed
==============
+ wirecardpaymentgateway/models/PaymentAlipayCrossborder.php  -1
+ wirecardpaymentgateway/models/PaymentSepaDirectDebit.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405&bid=15630647)

File: wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php (hunk 1 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Hook/BeforeOrderStatusUpdateHandler.php]: Codacy Issue found: [Avoid unused private methods such as 'isForceOrderStateChangeAllowed'.](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

File: wirecardpaymentgateway/classes/Constants/FormConstants.php (hunk 2 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Constants/FormConstants.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 25
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/Failure.php  1
- wirecardpaymentgateway/helper/StringHelper.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/EntityBuilderFactory.php  2
- wirecardpaymentgateway/helper/ArrayHelper.php  3
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/Basket/BasketBuilder.php  2
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/InitialTransactionBuilder.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItemCollection.php  3
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItem.php  3
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/TransactionCart.php  1
- wirecardpaymentgateway/helper/PaymentProvider.php  2
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/PostProcessingTransactionBuilder.php  2
- wirecardpaymentgateway/models/PaymentSofort.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -1
+ wirecardpaymentgateway/controllers/admin/WirecardTransactions.php  -10
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php  -3
         

Clones removed
==============
+ wirecardpaymentgateway/models/PaymentAlipayCrossborder.php  -1
+ wirecardpaymentgateway/models/PaymentSepaDirectDebit.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405&bid=15630647)

File: wirecardpaymentgateway/classes/Constants/FormConstants.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
```
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Constants/FormConstants.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 25
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/TransactionCart.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItemCollection.php  3
- wirecardpaymentgateway/classes/Transaction/Entity/Cart/CartItem.php  3
- wirecardpaymentgateway/helper/PaymentProvider.php  2
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/PostProcessingTransactionBuilder.php  2
- wirecardpaymentgateway/helper/StringHelper.php  1
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/EntityBuilderFactory.php  2
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/Entity/Basket/BasketBuilder.php  2
- wirecardpaymentgateway/classes/Transaction/Builder/InitialTransactionBuilder.php  1
- wirecardpaymentgateway/models/PaymentSofort.php  1
- wirecardpaymentgateway/helper/ArrayHelper.php  3
- wirecardpaymentgateway/classes/Response/Failure.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -1
+ wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php  -3
+ wirecardpaymentgateway/controllers/admin/WirecardTransactions.php  -10
         

Clones removed
==============
+ wirecardpaymentgateway/models/PaymentAlipayCrossborder.php  -1
+ wirecardpaymentgateway/models/PaymentSepaDirectDebit.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405&bid=15630647)

File: wirecardpaymentgateway/classes/Constants/FormConstants.php (hunk 2 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Constants/FormConstants.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4670405)

**Comment by tatsta** [comment]
Could you please rename this branch to RC-2.5.0-minor ? UI tests should start automatically then

**Pull request closed without merging**